server/src/server.rs

```diff
@@ -0,0 +1,709 @@
+		f.debug_struct("Server")
+			.field("listener", &self.listener)
+			.field("cfg", &self.cfg)
+			.field("id_provider", &self.id_provider)
+			.field("resources", &self.resources)
+			.finish()
+	}
+}
+
+impl<B, L> Server<B, L> {
+	/// Returns socket address to which the server is bound.
+	pub fn local_addr(&self) -> Result<SocketAddr, Error> {
+		self.listener.local_addr().map_err(Into::into)
+	}
+}
+
+impl<B, U, L> Server<B, L>
+where
+	L: Logger,
+	B: Layer<TowerService<L>> + Send + 'static,
+	<B as Layer<TowerService<L>>>::Service: Send
+		+ Service<
+			hyper::Request<hyper::Body>,
+			Response = hyper::Response<U>,
+			Error = Box<(dyn StdError + Send + Sync + 'static)>,
+		>,
+	<<B as Layer<TowerService<L>>>::Service as Service<hyper::Request<hyper::Body>>>::Future: Send,
+	U: HttpBody + Send + 'static,
+	<U as HttpBody>::Error: Send + Sync + StdError,
+	<U as HttpBody>::Data: Send,
+{
+	/// Start responding to connections requests.
+	///
+	/// This will run on the tokio runtime until the server is stopped or the `ServerHandle` is dropped.
+	pub fn start(mut self, methods: impl Into<Methods>) -> Result<ServerHandle, Error> {
+		let methods = methods.into().initialize_resources(&self.resources)?;
+		let (stop_tx, stop_rx) = watch::channel(());
+
+		let stop_handle = StopHandle::new(stop_rx);
+
+		match self.cfg.tokio_runtime.take() {
+			Some(rt) => rt.spawn(self.start_inner(methods, stop_handle)),
+			None => tokio::spawn(self.start_inner(methods, stop_handle)),
+		};
+
+		Ok(ServerHandle::new(stop_tx))
+	}
+
+	async fn start_inner(self, methods: Methods, stop_handle: StopHandle) {
+		let max_request_body_size = self.cfg.max_request_body_size;
+		let max_response_body_size = self.cfg.max_response_body_size;
+		let max_log_length = self.cfg.max_log_length;
+		let allow_hosts = self.cfg.allow_hosts;
+		let resources = self.resources;
+		let logger = self.logger;
+		let batch_requests_supported = self.cfg.batch_requests_supported;
+		let id_provider = self.id_provider;
+		let max_subscriptions_per_connection = self.cfg.max_subscriptions_per_connection;
+
+		let mut id: u32 = 0;
+		let connection_guard = ConnectionGuard::new(self.cfg.max_connections as usize);
+		let mut connections = FutureDriver::default();
+		let mut incoming = Monitored::new(Incoming(self.listener), &stop_handle);
+
+		loop {
+			match connections.select_with(&mut incoming).await {
+				Ok((socket, _addr)) => {
+					if let Err(e) = socket.set_nodelay(true) {
+						tracing::warn!("Could not set NODELAY on socket: {:?}", e);
+						continue;
+					}
+
+					let conn = match connection_guard.try_acquire() {
+						Some(conn) => conn,
+						None => {
+							tracing::warn!("Too many connections. Please try again later.");
+							connections.add(http::reject_connection(socket).boxed());
+							continue;
+						}
+					};
+
+					let tower_service = TowerService {
+						inner: ServiceData {
+							remote_addr: socket.peer_addr().ok(),
```

Comment:
this is why we have `Option<SocketAddr>` in the logging not sure how hyper can have SocketAddr

perhaps it's more or less fine to unwrap here?!

Reply:
I couldn't say offhand without looking into the `std` code to see why it returns a result; maybe worth a little dig to see if it's actually ok to unwrap?